### ISSUE TITLE
Supporting signal extended value type list

### ISSUE THREAD

**Issue by byunghun-ee** [xR3b0rn/dbcppp#2]
Hello xR3b0rn,

I have added a feature to parse "SIG_VALTYPE_" section for Signals with value types 'float' and 'double'.
I would like to contribute your dbc parser if you allow me contribute dbcppp.

There is no way to contact you directly. So, I just report an issue.

Thank you.

**Comment by xR3b0rn** [comment]
I appreciate contributions. Do you want create a pull request? Then I can merge your code into the master.

**Comment by ByunghunLee** [comment]
Yes, I will.

Thank you.

**Comment by xR3b0rn** [comment]
Can this be closed?

PS.: do you were intended to provide your test DBC-file (D:/sample.dbc)?

**Comment by ByunghunLee** [comment]
Yes, you can close this issue.

Unfortunately, I cannot provide you of the dbc file.
But If you give me your dbc file, I will add SIG_VALTYPE_ section.

Thank you.

**Comment by xR3b0rn** [comment]
I introduced serialization of the network back to DBC as new feature for simplifiing the process of verifiing the correctness of the parser. Therefore I changed some lines of your code:
- put information about extended value types into SignalImpl.
- introduced new enum: `class ExtendedValueType { Signed, Float, Double};`
- changed the type of `_extended_value_type` to `boost::optional<Signal::ExtendedValueType>`

**Comment by ByunghunLee** [comment]
Hello xR3b0rn,

I reviewed your commits.
If you want to serialize a network to a DBC file, I think _SignalExtendedValueType_ would be required as well since SIG_VALTYPE_ is network level like attributes. 

I have attached a modified _Model3CAN_ dbc file, which have new messages and signals whose types are float and double.

_SIG_VALTYPE_ 1538 signal_double : 2;_
_SIG_VALTYPE_ 1537 signal_float2 : 1;_
_SIG_VALTYPE_ 1537 signal_float1 : 1;_

Float and Double are decoded differently from Integers.

If you have any question, please let me know.
I am looking forward to your responds.

[Model3CAN_float_double.txt](https://github.com/xR3b0rn/dbcppp/files/4271113/Model3CAN_float_double.txt)

**Comment by xR3b0rn** [comment]
In my opinion the less relations in the network the better. Therefore I put the `SIG_VALTYPE` into the `Signal`. I also don't think there is any meaningful point in having a `SIG_VALTYPE` without having the associated `Signal`.

If I overlook something, feel free to argue.

> Float and Double are decoded differently from Integers.
The `SIG_VALTYPE` changes interpretion of the bit data received?
What is happening when `bit_size` != 32 but _extended_value_type == Foat?

**Comment by ByunghunLee** [comment]
I just tested it from Vector tool.
The bit sizes for float and double are fixed 32bit and 64bit, respectively.

If the signal value is float or double, they use IEEE 754.

      Sign  Exponent     Fraction 
       0      00000000   00000000000000000000000
Bit  31    [30 - 23]       [22        -         0]

**Comment by xR3b0rn** [comment]
This should speed up deocoding in for those kind of values in controst to nomal values. But I guess it's a bit of work, I try to create a pull request these days.

**Comment by xR3b0rn** [comment]
https://github.com/xR3b0rn/dbcppp/blob/feature_decode_ExtendedValueType/src/dbcppp/SignalImpl.cpp

**Comment by ByunghunLee** [comment]
Hello xR3b0rn,

I found an issue when loading multiple dbcs , which are involved in a
network.
Previously, dbcppp was able to load multiple dbcs.

In practice, each message is defined in separated dbc even they are in a
network.
Do you intend to load only one dbc which includes all network information?

Thank you.
Regards






2020년 2월 29일 (토) 오후 8:24, xR3b0rn <notifications@github.com>님이 작성:

**Comment by xR3b0rn** [comment]
Thanks for your report.
The issue is caused by the internal data structure change.

For supporting mering two different DBCs the API has to be adapted. This shouldn't be a huge problem. I add this as TODO.

**Comment by xR3b0rn** [comment]
It's now possible to "merge" two DBCs by using `std::unique_ptr<Network> Network::fromDBC(std::istream& is, std::unique_network<Network> network);`. For future request, please create new issues.

**Comment by ByunghunLee** [comment]
Thank you very much. I will try it, then let you know.
I reported an issue on parsing bit timing when bit timing is empty.

Question 1.
Do you only consider supporting CAN and CANFD bus?
The reason I ask this question, DBC files are used for parsing serial
communication protocol with same rule of parsing and decoding. However, the
message size is dynamic.

2020년 3월 3일 (화) 오전 7:03, xR3b0rn <notifications@github.com>님이 작성:

**Comment by xR3b0rn** [comment]
Can you calrify this usecase? CANFD is dynamic sized, but does have at most 64 bytes, so you have always pass a 64 byte array to the decode64-function. Which kind of protocols you are talking about?

**Comment by xR3b0rn** [comment]
I thought about it. I think it should be possible to create a decode function without fixed data sizes when taking the message size into account. I create an issue for this.

**Comment by ByunghunLee** [comment]
After huge change, Parser throws alot of errors.
Also, your test DBC is not opened by Canoe.

So, I attached a sample DBC file. Please understand that I changed name of
message and signals.

Please try it. Also, I will provide you of another DBC including FLOAT
signal.

Thank you.
Regards

2020년 3월 3일 (화) 오후 4:23, xR3b0rn <notifications@github.com>님이 작성:

**Comment by xR3b0rn** [comment]
Where can I find your DBC?

**Comment by ByunghunLee** [comment]
I have attached tesA.dbc in previous mail.
Please find it.

Thank you.

2020년 3월 4일 (수) 오전 8:58, xR3b0rn <notifications@github.com>님이 작성:

**Comment by xR3b0rn** [comment]
I can't see anything...

**Comment by ByunghunLee** [comment]
please try this link.

Although the message size is 1940 bytes, actual data size varies depending
on the condition.


https://drive.google.com/open?id=1Mp2_OIiFf06qCQuizIOrM3ot97xHrTEG

One issue is that after NS_, only when one space allow the parser parses
the file. For example, "tab(/t)" causes issue.

Thank you.

2020년 3월 4일 (수) 오전 9:43, xR3b0rn <notifications@github.com>님이 작성:

**Comment by xR3b0rn** [comment]
This file parses without any problems. I'm aware of the issue you addressed, I will add this as an issue.

**Comment by ByunghunLee** [comment]
Please find the following link which is a log of dvcppp_test with the dbc.
https://drive.google.com/file/d/1QoRLgnSBVyz-G8XcbSXS_8UT1phK3aDm/view?usp=sharing

I tested this in Windows10 x64 system.

Thank you.

2020년 3월 4일 (수) 오전 10:59, xR3b0rn <notifications@github.com>님이 작성:

**Comment by xR3b0rn** [comment]
Yeah, this is what I'm tlaking about.

**Comment by xR3b0rn** [comment]
The
> One issue is that after NS_, only when one space allow the parser parses
the file. For example, "tab(/t)" causes issue.
issue is fixed.

**Comment by xR3b0rn** [comment]
Decoding of nbytes-messages is now supported.
Now only one decoding function exists in which you can pass a byte array with arbitrarily size to the function.

**Comment by ByunghunLee** [comment]
Windows 10 x64 with Clion + vs2017 compiler

To uses std::function,
#include <functional> is required.

Thank you.


2020년 3월 7일 (토) 오전 7:18, xR3b0rn <notifications@github.com>님이 작성:

**Comment by xR3b0rn** [comment]
Thank you for the report.

Please open new issues tagged with bug the next time. Because then I'm able to link the commit with the issue.

**Comment by ByunghunLee** [comment]
In order to load two separated dbcs into network, I have tried to fromDBC
methods.

auto net = dbcppp::Network::fromDBC(idbc);
dbcppp::Network::fromDBC(idbc1, std::move(net));

When I tested it, it is shown that "fatal error: in "DBCParsing": DBC
parsing failed!" from "BOOST_REQUIRE_MESSAGE(net, "DBC parsing
failed!");"

Is this right way to load multiple DBCs?

Thank you.



2020년 3월 11일 (수) 오후 2:09, angelo lee <byunghun.ee@gmail.com>님이 작성:
